test/manipulation.test.js

```diff
@@ -91,6 +118,20 @@ describe('manipulation', function () {
       });
     });
 
+    it('should not allow user-defined value for the id of object - create (promise variant)', function (done) {
+      Person.create({id: 123456})
+      .then (function (p) {
+        should.not.exist(p);
+        done();
+      })
+      .catch(function (err) {
+        err.should.be.instanceof(ValidationError);
+        err.statusCode.should.equal(422);
+        err.details.messages.id.should.eql(['can\'t be set']);
+        done();
+      });
```

Comment:
catch(function (err) {...}) is basically syntactic sugar for then(null,
(err) {...}).
In this case, your second catch statement wouldn't do anything unless
another error is thrown from the first error handler, which it isn't.

In general, I like to use catch rather than a second argument to then
because it's more visible than just a comma and thus easier to see its
intent as an error handler.

On Thu, Feb 19, 2015 at 10:12 AM, Miroslav Bajtoš <notifications@github.com>
wrote:

> In test/manipulation.test.js
> <https://github.com/strongloop/loopback-datasource-juggler/pull/451#discussion_r25005283>
> :
>
> > @@ -91,6 +118,20 @@ describe('manipulation', function () {
> >        });
> >      });
> >
> > +    it('should not allow user-defined value for the id of object - create (promise variant)', function (done) {
> > +      Person.create({id: 123456})
> > +      .then (function (p) {
> > +        should.not.exist(p);
> > +        done();
> > +      })
> > +      .catch(function (err) {
> > +        err.should.be.instanceof(ValidationError);
> > +        err.statusCode.should.equal(422);
> > +        err.details.messages.id.should.eql(['can\'t be set']);
> > +        done();
> > +      });
>
> This is weird. Are you expecting that Person.create succeeds or fails?
> Looking at the test above, the operation should fail. In which case you
> need to assert in the then block that the operation should have failed.
>
> Person.create({ id: 123456 })
>   .then(function(p) {
>     return done(new Error('Operation should have failed.'));
>   })
>   .catch(function(err) {
>     // ...
>   });
>
> Also have you seen this test fail? AFAIK, catch converts any errors
> thrown from the handler into a promise rejection, thus you need to add one
> more catch handler at the end. And since then accepts both success and
> failure handlers,
> the code can look like this:
>
> Person.create({ id: 123456 })
>   .then(
>     function(p) { done(new Error('Person.create should have failed.')); },
>     function(err) {
>       // asserts
>     })
>   .catch(done);
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/strongloop/loopback-datasource-juggler/pull/451/files#r25005283>
> .
>